### ISSUE TITLE
PID conflict with logging module

### ISSUE THREAD

**Issue by leilerg** [trbs/pid#31]
I've been working on a python daemon with `pid`, and after initial logs directly to console I wanted to switch to file logging, using python's `logging` module. This is when I run into the problem. 

I have `start`/`stop` functions to manage the daemon:

```
import os
import sys
import time
import signal
import lockfile
import logging
import logging.config

import daemon
from pid import PidFile

from mpmonitor.monitor import MempoolMonitor

# logging.config.fileConfig(fname="logging.conf", disable_existing_loggers=False)
# log = logging.getLogger("mpmonitor")


def start():
    print("Starting Mempool Monitor")

    _pid_file = PidFile(pidname="mpmonitor.pid", piddir=curr_dir)

    with daemon.DaemonContext(stdout=sys.stdout,
                              stderr=sys.stderr,
                              stdin=sys.stdin,
                              pidfile=_pid_file):

        # Start the monitor:
        mpmonitor = MempoolMonitor()
        mpmonitor.run()
    

def stop():
    print("\n{}\n".format(pid_file))

    try:
        with open(pid_file, "r") as f:
            content = f.read()
        f.close()

    except FileNotFoundError as fnf_err:
        print("WARNING - PID file not found, cannot stop daemon.\n({})".format(pid_file))
        sys.exit()


    print("Stopping Mempool Monitor")
    # log.info("Stopping Mempool Monitor")
    pid = int(content)
    os.kill(pid, signal.SIGTERM)

    sys.exit()
```

which works as you would expect it to.

&nbsp;

If I uncomment the logging code, 
```
logging.config.fileConfig(fname="logging.conf", disable_existing_loggers=False)
log = logging.getLogger("mpmonitor")
```
the code breaks down, and some pretty random stuff happens. The error message:

```
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.6/logging/__init__.py", line 998, in emit
    self.flush()
  File "/usr/lib/python3.6/logging/__init__.py", line 978, in flush
    self.stream.flush()
OSError: [Errno 9] Bad file descriptor
Call stack:
  File "mpmonitor.py", line 111, in <module>
    start()
  File "mpmonitor.py", line 48, in start
    pidfile=_pid_file):
  File "/home/leilerg/.local/lib/python3.6/site-packages/daemon/daemon.py", line 389, in __enter__
    self.open()
  File "/home/leilerg/.local/lib/python3.6/site-packages/daemon/daemon.py", line 381, in open
    self.pidfile.__enter__()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 216, in __enter__
    self.create()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 168, in create
    self.setup()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 77, in setup
    self.logger.debug("%r entering setup", self)
Message: '%r entering setup'
Arguments: (<pid.PidFile object at 0x7fc8faa479e8>,)
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.6/logging/__init__.py", line 998, in emit
    self.flush()
  File "/usr/lib/python3.6/logging/__init__.py", line 978, in flush
    self.stream.flush()
OSError: [Errno 9] Bad file descriptor
Call stack:
  File "mpmonitor.py", line 111, in <module>
    start()
  File "mpmonitor.py", line 48, in start
    pidfile=_pid_file):
  File "/home/leilerg/.local/lib/python3.6/site-packages/daemon/daemon.py", line 389, in __enter__
    self.open()
  File "/home/leilerg/.local/lib/python3.6/site-packages/daemon/daemon.py", line 381, in open
    self.pidfile.__enter__()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 216, in __enter__
    self.create()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 170, in create
    self.logger.debug("%r create pidfile: %s", self, self.filename)
Message: '%r create pidfile: %s'
Arguments: (<pid.PidFile object at 0x7fc8faa479e8>, '/home/leilerg/python/mempoolmon/mpmonitor.pid')
Traceback (most recent call last):
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 136, in inner_check
    pid = int(pid_str)
ValueError: invalid literal for int() with base 10: 'DEBUG - 2020-04-'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "mpmonitor.py", line 111, in <module>
    start()
  File "mpmonitor.py", line 48, in start
    pidfile=_pid_file):
  File "/home/leilerg/.local/lib/python3.6/site-packages/daemon/daemon.py", line 389, in __enter__
    self.open()
  File "/home/leilerg/.local/lib/python3.6/site-packages/daemon/daemon.py", line 381, in open
    self.pidfile.__enter__()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 216, in __enter__
    self.create()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 180, in create
    check_result = self.check()
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 165, in check
    return inner_check(self.fh)
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 139, in inner_check
    raise PidFileUnreadableError(exc)
pid.PidFileUnreadableError: invalid literal for int() with base 10: 'DEBUG - 2020-04-'
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.6/logging/__init__.py", line 998, in emit
    self.flush()
  File "/usr/lib/python3.6/logging/__init__.py", line 978, in flush
    self.stream.flush()
OSError: [Errno 9] Bad file descriptor
Call stack:
  File "/home/leilerg/.local/lib/python3.6/site-packages/pid/__init__.py", line 197, in close
    self.logger.debug("%r closing pidfile: %s", self, self.filename)
Message: '%r closing pidfile: %s'
Arguments: (<pid.PidFile object at 0x7fc8faa479e8>, '/home/leilerg/python/mempoolmon/mpmonitor.pid')
```

&nbsp;

The random stuff I was referring to is now the file `mpmonitor.pid` doesn't contain a PID anymore but some attempted logs/error messages

```
user@mylaptor:mempoolmon: cat mpmonitor.pid

DEBUG - 2020-04-05 10:52:55,676 - PidFile: <pid.PidFile object at 0x7fc8faa479e8> entering setup
DEBUG - 2020-04-05 10:52:55,678 - PidFile: <pid.PidFile object at 0x7fc8faa479e8> create pidfile: /home/leilerg/python/mempoolmon/mpmonitor.pid
DEBUG - 2020-04-05 10:52:55,678 - PidFile: <pid.PidFile object at 0x7fc8faa479e8> check pidfile: /home/leilerg/python/mempoolmon/mpmonitor.pid
DEBUG - 2020-04-05 10:52:55,678 - PidFile: <pid.PidFile object at 0x7fc8faa479e8> closing pidfile: /home/leilerg/python/mempoolmon/mpmonitor.pid
```

&nbsp;

To me this looks like the `pid` logfile got confused with the PID file somehow. This is odd as I explicitly set `disable_existing_loggers=False`.

Any ideas? Am I doing something not-so-obviously wrong?

**Comment by trbs** [comment]
this is weird..

I tried to reproduce this but had no luck.
Here is my attempt.

**test_issue_31.py**
```python
#!/usr/bin/env python3

import os
import sys
import time
import signal
import logging
import logging.config

import daemon
from pid import PidFile

logging.config.fileConfig(fname="logging.conf", disable_existing_loggers=False)
log = logging.getLogger("test_issue_31")

PIDNAME = "/tmp/test_issue_31.pid"


def start():
    print(f"Starting {PIDNAME}")

    _pid_file = PidFile(pidname=PIDNAME)

    with daemon.DaemonContext(stdout=sys.stdout,
                              stderr=sys.stderr,
                              stdin=sys.stdin,
                              pidfile=_pid_file):

        while 1:
            time.sleep(0.1)


def stop():
    print("\n{}\n".format(PIDNAME))

    try:
        with open(PIDNAME, "r") as f:
            content = f.read()
        f.close()
    except FileNotFoundError as fnf_err:
        print(fnf_err)
        print("WARNING - PID file not found, cannot stop daemon.\n({})".format(PIDNAME))
        sys.exit()

    print("Stopping")
    # log.info("Stopping Mempool Monitor")
    pid = int(content)
    os.kill(pid, signal.SIGTERM)

    sys.exit()


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("specify 'start' or 'stop'")
    elif sys.argv[1] == "start":
        start()
    else:
        stop()
```

**logging.conf**
```
[loggers]
keys=root

[handlers]
keys=stream_handler

[formatters]
keys=formatter

[logger_root]
level=DEBUG
handlers=stream_handler

[handler_stream_handler]
class=StreamHandler
level=DEBUG
formatter=formatter
args=(sys.stderr,)

[formatter_formatter]
format=%(asctime)s %(name)-12s %(levelname)-8s %(message)s
```

**Execution:**
```
$ PYTHONPATH=. ./test_issue_31.py start
Starting /tmp/test_issue_31.pid
2020-04-05 18:20:53,948 PidFile      DEBUG    <pid.posix.PidFile object at 0x10dd3cbb0> entering setup
2020-04-05 18:20:53,949 PidFile      DEBUG    <pid.posix.PidFile object at 0x10dd3cbb0> create pidfile: /tmp/test_issue_31.pid
2020-04-05 18:20:53,950 PidFile      DEBUG    <pid.posix.PidFile object at 0x10dd3cbb0> check pidfile: /tmp/test_issue_31.pid
```
```
$ cat /tmp/test_issue_31.pid
10204
```
```
$ PYTHONPATH=. ./test_issue_31.py stop
/tmp/test_issue_31.pid
Stopping
2020-04-05 18:21:11,749 PidFile      DEBUG    <pid.posix.PidFile object at 0x10dd3cbb0> closing pidfile: /tmp/test_issue_31.pid
Terminating on signal 15
2020-04-05 18:21:11,749 PidFile      DEBUG    <pid.posix.PidFile object at 0x10dd3cbb0> closing pidfile: /tmp/test_issue_31.pid
```
```
$ cat /tmp/test_issue_31.pid
cat: /tmp/test_issue_31.pid: No such file or directory
```

**Comment by leilerg** [comment]
Running your code without problems, but this logs to console, not file. If I try to log to file, it fails as above.

###logging.conf###

```
[loggers]
keys=root


[handlers]
keys=fileHandler


[formatters]
keys=basicFormatter


[logger_root]
level=NOTSET
handlers=fileHandler


[handler_fileHandler]
level=DEBUG
class=FileHandler
formatter=basicFormatter
args=("mpmonitor.log", "a")


[formatter_basicFormatter]
format=%(levelname)s - %(asctime)s - %(name)s: %(message)s
```

&nbsp;

Can you see any problems with my `logging.conf`?

**Comment by trbs** [comment]
This is very interesting...

I get the same issue when using `class=FileHandler`.
Don't see anything obviously wrong with either the config nor `pidfile`..

When removing `with daemon.DaemonContext(...):` it works as expected. 

I think the reason for this is the following.

What's happening here is kinda interesting and I think it's not really a bug in `pidfile` itself but it's the first victim of it.

Pidfile when opening the pid file will open a file descriptor with number 3.

- File descriptor 0 is stdin
- File descriptor 1 is stdout
- File descriptor 2 is stderr
- File descriptor 3 is /tmp/test_issue_31.pid

This is inside of the process that `DaemonContext` forked.

However the main process already opened file descriptor 3 when logging creating the log file.
Thus unaware of each other they are writing to the same file.

A further explanation can be found here: https://stackoverflow.com/questions/13180720/maintaining-logging-and-or-stdout-stderr-in-python-daemon

Therefor the solution is probably something like this:

**test_issue_31.py**
```
#!/usr/bin/env python3

import os
import sys
import time
import signal
import logging
import logging.config

import daemon
from pid import PidFile

logging.config.fileConfig(fname="logging.conf", disable_existing_loggers=False)
log = logging.getLogger("test_issue_31")

PIDNAME = "/tmp/test_issue_31.pid"


def get_logging_handles(logger):
    handles = []
    for handler in logger.handlers:
        handles.append(handler.stream.fileno())
    if logger.parent:
        handles += get_logging_handles(logger.parent)
    return handles


def start():
    print(f"Starting {PIDNAME}")

    files_preserve = get_logging_handles(logging.root)
    _pid_file = PidFile(pidname=PIDNAME)

    with daemon.DaemonContext(stdout=sys.stdout,
                              stderr=sys.stderr,
                              stdin=sys.stdin,
                              pidfile=_pid_file,
                              files_preserve=files_preserve):
        while 1:
            time.sleep(0.1)
    print("DONE")


def stop():
    print("\n{}\n".format(PIDNAME))

    try:
        with open(PIDNAME, "r") as f:
            content = f.read()
        f.close()
    except FileNotFoundError as fnf_err:
        print(fnf_err)
        print("WARNING - PID file not found, cannot stop daemon.\n({})".format(PIDNAME))
        sys.exit()

    print("Stopping")
    # log.info("Stopping Mempool Monitor")
    pid = int(content)
    os.kill(pid, signal.SIGTERM)

    sys.exit()


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("specify 'start' or 'stop'")
    elif sys.argv[1] == "start":
        start()
    else:
        stop()
```

**Comment by trbs** [comment]
If so... please please write a PR with a little documentation for in the readme or alike that this might help other people in the future :-)

**Comment by leilerg** [comment]
OK, this works with a caveat. I need to `sudo` the code as without it I'm getting 

```
 File ".local/lib/python3.6/site-packages/pid/__init__.py", line 107, in _make_filename
    raise IOError("Pid file directory '%s' cannot be written to" % piddir)
OSError: Pid file directory '/var/run/' cannot be written to
```

Do you get the same? I can survive with it, but ideally I'd like to run this as any user.

&nbsp;

I'm happy to write a documentation PR, but bear with me as I'm a total noob with FOSS projects, so will take some time to figure it out.

**Comment by trbs** [comment]
Could you show us the permissions on `/var/run` ?

Normally if the directory `/var/run` is not writable it should should fallback on the default temp dir location of the system. (normally `/tmp`)

But you could always explicitly set `piddir` on `PidFile`.

```python
_pid_file = PidFile(pidname=PIDNAME, piddir="/tmp")
```

**Comment by leilerg** [comment]
```
$ ls -la /var/run 

lrwxrwxrwx 1 root root 4 Oct 27  2018 /var/run -> /run


$ ls -al /run

total 36
drwxr-xr-x 34 root  root  1040 Mar 17 14:40 .
drwxr-xr-x 25 root  root  4096 Mar  3 18:18 ..
```

Clearly only `root` has write access to `/run`, but I don't know if this should be any different. These are default settings for Linux Mint... don't think I've ever touched these.

&nbsp;

```
_pid_file = PidFile(pidname=PIDNAME, piddir="/tmp")
```

Works!

**Comment by trbs** [comment]
Weird the check on write permissions should catch this... so I think we got ourselves a little bug here ;-)

Released 3.0.1 that should fix that...

For now I think I can close this ticket right ?
Seems you got everything working :-)
